provider/src/cache.ts

```diff
@@ -151,3 +82,129 @@ export const storeContractStatusInKV = async (
     console.error('Error storing contract status in KV:', error)
   }
 }
+
+/**
+ * Get a cached response from KV storage.
+ * It retrieves and parses the stored value as JSON.
+ * If reading or parsing fails, it attempts to delete the corrupted cache entry and returns null.
+ * @param request The RPC request
+ * @param chainId The chain ID
+ * @param env The environment variables
+ * @returns The cached response or null if not found
+ */
+export const getResponseFromKV = async (
+  request: RpcRequest,
+  chainId: string,
+  env: Env
+): Promise<any> => {
+  if (!env.REQUEST_CACHE) {
+    return null
+  }
+
+  const cacheKey = generateRequestCacheKey(request)
+
+  try {
+    const cached = await env.REQUEST_CACHE.get(cacheKey, 'json')
+    if (cached) {
+      return cached
+    }
+  } catch (error) {
+    console.error(
+      `Error reading from cache: ${error instanceof Error ? error.message : String(error)}`
+    )
+    try {
+      await env.REQUEST_CACHE.delete(cacheKey)
+    } catch (err) {
+      console.error(`Error deleting corrupted cache entry: ${err}`)
+    }
+  }
+
+  return null
+}
+
+/**
+ * Cache a response
+ * @param request The RPC request
+ * @param response The response to cache
+ * @param chainId The chain ID
+ * @param env The environment variables
+ */
+export const storeResponseInKV = async (
+  request: RpcRequest,
+  response: any,
+  chainId: string,
+  env: Env
+): Promise<void> => {
+  if (!env.REQUEST_CACHE || !response?.result) {
+    return
+  }
+  const cacheKey = generateRequestCacheKey(request)
+  const ttl = getCacheTTL(env)
+
+  try {
+    const cacheValue = { result: response.result }
+    await env.REQUEST_CACHE.put(cacheKey, JSON.stringify(cacheValue), {
```

Comment:
does this work? I assume if we can "Read" json, it should store json?
```suggestion
    await env.REQUEST_CACHE.put(cacheKey, cacheValue, {
```

Reply:
no it doesn't. the `put` method only accepts strings, or an array buffer